_posts/2018-08-30-on-the-fly-transformations.md

```diff
@@ -0,0 +1,342 @@
+needed for that particular analysis.
+Trajectory transformations in MDAnalysis, on the other end, have one great advantage - they
+are performed on-the-fly for each frame that is read. Transformations are added to a universe
+as a transformation workflow containing one or more transformations. The API also makes it easy
+to add new transformations for your own projects.
+Another things that really makes the "on-the-fly" aspect of the MDAnalysis transformations
+shine is coupling it to a visualization widget such as
+[NGL Viewer](http://nglviewer.org/ngl/api/index.html).
+
+## Using MDAnalysis transformations
+Now it's time to learn how to use the trajectory transformations in MDAnalysis. During the
+following steps, we will apply some transformations on a 1 ns trajectory of a simple
+19-residue peptide embeded in a 128-DMPC membrane, showing the GROMACS `trjconv` command
+and the equivalent MDAnalysis code and output. To keep thinks lightweight, frames are
+were taken every 100 ps, and water molecules were removed. This can be easily done with
+MDAnalysis.
+
+### Example 1: making everything whole again
+When performing MD simulations using periodic boundary conditions, molecules will often
+cross the limits of the unit cell. When this happens, some atoms of the molecule will
+show up on the the opposing side of the unit cell and some molecular viewers will show
+stretched bonds and other visual artifacts depending on the visual representation of the
+system. 
+This is the case of our system. Without any modifications, when we look at the trajectory
+of our system, things become more cluttered and confusing:
+
+
+```python
+import warnings
+warnings.filterwarnings('ignore') # some attributes are missing 
+import MDAnalysis as mda
+import nglview as nv
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+nv.show_mdanalysis(u)
+```
+
+
+Using `trjconv`, one way to may every molecules whole again would be:
+
+    gmx trjconv -f pept_in_memb.xtc -s pept_in_memb.tpr -pbc mol -o output.xtc
+    
+In MDAnalysis this can be done with the `unwrap` transformation, which takes an AtomGroup as argument.
+This can be done as follows:
+
+
+```python
+# a custom atom group can be passed as an argument. In this case we will use all the atoms
+# in the Universe u
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+ag = u.atoms
+# we define the transformation
+transformations = mda.transformations.unwrap(ag)
+```
+
+Now that we have a workflow - in this case it is only a single transformation - we add
+it to the `trajectory` object so it can be applied in each frame that we want to read.
+
+```python
+u.trajectory.add_transformations(transformations)
+```
+
+If we want to, we can do other things with the trajectory without having to generate a new file
+with the transformed trajectory.
+
+This is how our trajectory looks like:
+
+```python
+nv.show_mdanalysis(u)
+```
```

Comment:
@lilyminium would there be a way to embed the nglview widget here with actual content — it would be neat if users could see the changes.

Reply:
If you really wanted to, I think you could do this (@hainm would know more):
1. Run the code in a Jupyter notebook and write view out to an HTML file
```
view = nv.show_mdanalysis(u)
view  # you must visualise it in the browser
view.write_html('transformed.html', [view], frame_range=[0, len(u), 1])
```
2. Scrap everything outside the `<body>` tags in `transformed.html`
3. `{% include transformed.html %}` in your post

Otherwise making it a gif and including the gif is probably easier.

Reply:
I have created the html files. Do I simply paste the code inside the `<body>` tags in the md file?

Reply:
You could, but I think that using the [include](https://jekyllrb.com/docs/includes/) tag would be functionally the same and neater.

I think you could also make an ``iframe.html`` that looks like this, stick it in `_includes/`

``<iframe src={{ include.nv_html }}> </iframe>``

and in your .md file go:

`{% include iframe.html nv_html="transformed.html" %}`

which hopefully would remove the need to delete the stuff outside `<body />`.

Reply:
If that doesn't render well, here is an example of [using the movie maker to make a gif](https://lilyminium.github.io/UserGuide/examples/analysis/alignment_and_rms/aligning_trajectory.html#Writing-trajectories-to-a-file) (scroll up a bit).